### ISSUE TITLE
OpenSeaMap

### ISSUE THREAD

**Issue by indus** [osm2vectortiles/osm2vectortiles#402]
Would be nice to adapt this to the data shown at http://www.openseamap.org/

Am I right that "only" [this file](https://github.com/osm2vectortiles/osm2vectortiles/blob/master/osm2vectortiles.tm2source/data.yml) has to be manipulated the right way to select data from within the **[:seamark](http://wiki.openstreetmap.org/wiki/Key:seamark)** namespace from OSM to pack them into vtiles?

Has somebody a [8+ core 16+ GB RAM maschine](https://github.com/osm2vectortiles/osm2vectortiles/blob/master/USAGE.md) and some time to help me with that?

**Comment by lukasmartinelli** [comment]
Perhaps it makes sense to have a separate vector tile set just for sea map data?

You also don't need to scale it across multiple hosts if you don't need the expensive queries we do and limit yourself to z13 - i did something similar here https://github.com/lukasmartinelli/osm-noise-pollution.

Or if we want that in OSM2VectorTiles you first have to pick the tags that you want in the map.
And then add those to the imposm3 mapping here:  https://github.com/osm2vectortiles/osm2vectortiles/blob/master/src/import-osm/mapping.yml

And then you can use it in the tm2source. (or use zoom level views in import-sql).

**Comment by aAXEe** [comment]
I am one of the developers of OpenSeaMap. We are currently working on a new map interface: https://github.com/aAXEe/online_chart_ol3
It uses OpenLayers3 and traditional raster tiles until now. But switching to vector tiles would be very nice.

I would prefer to add the seamark data in the main tiles .. the additional data is not that much as most of the seamarks are simple nodes with some tags. 
On the other hand we love to deliver map updates within minutes ... it is great for (osm) contributors to see their changes quickly on the map. Having a separate set can enable that, can't it?

We have a server for rendering tiles and other things: http://wiki.openseamap.org/wiki/OpenSeaMap-dev:Server_Bravo
I am sure you can get some resources on this machine.

The hard part of the map is the rendering of the symbols. There are tons of different symbols. Especially the lights are hard to get right: https://alpha.openseamap.org/#/?lat=54.1820&lon=12.0808&zoom=15&layers=A10001
Have a look at our current tile-renderer (written in C): https://github.com/OpenSeaMap/renderer/blob/master/searender/rules.c
As I am not familiar with the rendering I do not know how difficult it is to get this things rendered in OpenGL. Do you have an idea?

**Comment by lukasmartinelli** [comment]
> I am one of the developers of OpenSeaMap. We are currently working on a new map interface: https://github.com/aAXEe/online_chart_ol3
> It uses OpenLayers3 and traditional raster tiles until now. But switching to vector tiles would be very nice.

You can also use it just as baselayer.

> I would prefer to add the seamark data in the main tiles .. the additional data is not that much as most of the seamarks are simple nodes with some tags. 
On the other hand we love to deliver map updates within minutes ... it is great for (osm) contributors to see their changes quickly on the map. Having a separate set can enable that, can't it?

If you have a separate vector set and care about displaying immediate updates I suggest you render vector tiles on the fly.

Or if you store the features on low enough zoom levels (<10) you can also regenerate all tiles each hour.

> The hard part of the map is the rendering of the symbols. There are tons of different symbols. Especially the lights are hard to get right: https://alpha.openseamap.org/#/?lat=54.1820&lon=12.0808&zoom=15&layers=A10001
> Have a look at our current tile-renderer (written in C): https://github.com/OpenSeaMap/renderer/blob/master/searender/rules.c
> As I am not familiar with the rendering I do not know how difficult it is to get this things rendered in OpenGL. Do you have an idea?

But these symbols are quite important for you right? More complicated symbols are tricky (and you  will need to draw SVGs). If you put a lot of time into making the symbols this is a bit sad.

**Comment by indus** [comment]
Why is there a need for SVGs? I think mapbox-gl-js uses a single PNG spritesheet referenced by a property so as long as the symbols come with tranparancy, it should be easy to stich them in a grid-layout in a single PNG.
I would be happy to do that; it should be easier than any of other tasks to do for me.

The special sea-mal style-sheet makes me think that it would be better to keep "street" and "sea" seperate layers. (maybe it is possible to merge them on the fly when requested, like it is eine by mapbox with "street" and "terrain"?

**Comment by aAXEe** [comment]
I think the SVGs are ment to draw complex things like sectored lights: http://wiki.openstreetmap.org/wiki/Seamarks/Sectored_and_Directional_Lights

We will need some kind of dynamic picture for that.

You can find the existing SVG files here: https://github.com/OpenSeaMap/renderer/tree/master/searender/symbols

**Comment by indus** [comment]
Sorry. I thought you where talking about the various point symbols. Would it ne possible to split these big "symbols" into a bunch of polygons and polylines that could get rendered clientside without a predefined asset?
I would imagine it like a roundabout in the streets dataset: A point feature expanded to a complex shape.

**Comment by indus** [comment]
And now I also understand the problem with the point symbols. There is not a set of all possible symbols but your set contains fragments that get combined and colored  dynamically; e.g. pillars and barrels:
![image](https://cloud.githubusercontent.com/assets/2046274/17801620/785c3a6c-65eb-11e6-92ad-cc98ad1b445c.png)

**Comment by aAXEe** [comment]
>  There is not a set of all possible symbols but your set contains fragments that get combined and colored dynamically; e.g. pillars and barrels

Correct ;)
But these symbols can be combined from a predefined set. I think it should not be too hard to render this kind of things.

For the complex symbols like sectored lights: They are not like POIs or point symbols but like a street etc. They can cover a greater area. I thing converting them into real map objects is a better way. Something like @lukasmartinelli did with the noise map: https://github.com/lukasmartinelli/osm-noise-pollution

This would also fix a common problem: Parts from a light can be visible on the map while the point itself is not. Rendering on the client side would mean that the client has to look ahead of the current view.

**Comment by lukasmartinelli** [comment]
I set up a starting point on this repo for you here: https://github.com/osm2vectortiles/openseamap-skeleton. This should all suffice to generate vector tiles.

You will need to begin with defining the OSM tags you want in your database (well you could also import every seamark anyway but explicit makes it easier later to create good layers).

And then you can write the SQL queries in Mapbox Studio (also bundled in the skeleton).

If you want to work on your custom vector tiles we transfer that project over to your Open Sea Map organization and you can treat it as you like.

**Comment by indus** [comment]
I also set up a small [test](https://indus.github.io/OSeaM/index.html)
to get an idea what combining multiple icons actually means, and what the result could look like.
(I´m sorry to say, but I don´t have time to work on this at this weekend)

**Comment by aAXEe** [comment]
@lukasmartinelli: I copied the repo: https://github.com/OpenSeaMap/vectortiles-generator
I already added a simple "all seamarks" layer and exported the vector tiles.
Some issues outline the needed work. Is there something missing?

An instance of the tile server: http://vectortiles.195.37.132.70.xip.io/
I use only a small extract around our "map base" Warnemünde (DE) (It was the starting point of OpenSeaMap some years ago)
You can use http://vectortiles.195.37.132.70.xip.io/data/oseam.json as a tile source.

I updated my [example](https://aaxee.github.io/openseamap-gl/) to use the vector tiles from above and display them as red dots. 
(Note: the tile server is http only. In firefox you can disable the security warnings and see the map)

@indus: Your demo already looks great! Do you continue to work on this? Do you need something?

**Comment by indus** [comment]
@aAXEe Yes I'm going to continue this. Hard to if I will find much time during The week. But I will have at least oje das during The weekend I think. I think it would ne best to directly use your tile-server and concentrate on POIs first; what do you think?

**Comment by aAXEe** [comment]
>  I think it would be best to directly use your tile-server and concentrate on POIs first; what do you think?

Yes of course. I document the tile server here: https://github.com/OpenSeaMap/vectortiles-generator/wiki

Do you already have a repo for the style?
I would like to have it separated from the rest for the first time. It might be a good idea to add it to [OpenSeaMap/vectortiles-generator](https://github.com/OpenSeaMap/vectortiles-generator) later on because the style will only work with the tiles provided by this toolset (and settings).

I think [lukasmartinelli/osm-liberty](https://github.com/lukasmartinelli/osm-liberty) could be a good base for the base structure.

**Comment by indus** [comment]
@aAXEe I will continue with the [repo](https://github.com/indus/OSeaM) of the Demo. For now I will use your geojson; but without a property "category" it is hard to test varying styles of a type.

**Comment by aAXEe** [comment]
>For now I will use your geojson; but without a property "category" it is hard to test varying styles of a type.

I added a new layer for the vector tiles: `buoy_cardinal`
These entires have all attributes documented here: http://wiki.openstreetmap.org/wiki/Seamarks/Seamark_Objects

I tweaked the tile server ... you can now use the x-ray view from the tile server to see the available data: 
eg: http://vectortiles.195.37.132.70.xip.io/data/oseam/#13.22/54.1860/12.0490
The map is already centered at a cardinal buoy.

I am working on a system to automatically map all nodes documented in the osm wiki into the database and the vector tiles.

**Comment by indus** [comment]
I´ve updatedt my [Demo](https://indus.github.io/OSeaM/)

**Comment by aAXEe** [comment]
The tile server now has the complete seamark mapping available:
https://github.com/OpenSeaMap/vectortiles-generator/issues/1#issuecomment-242985393

**Comment by indus** [comment]
@aAXEe  I started a small test on bringing the current style together with the data you provide: https://indus.github.io/OSeaM/oseam/
It seems to me that the data has to be structured a bit different. In the moment stuff that is almost the same (from a style perspective ;-) is spread over multiple source-layers: e. g.various "pillar"  shapes in "buoy_lateral", "buoy_special_purpose", etc.

From my POV only the shape of a POI symbol matters, e.g. because it determinats the offset that is needed to place it in the right position and also to get the placement of combined Symbols correct; "colour" acts on top of "shape" as a modifier. It wouldnt change label placement for example. Splitting this property into "colour" and "colour-pattern" gives the need for a lot of duplicates in the style. A single modifier (e.g. "vyby" for vertical + yellow;black;yellow) would make things a lot easier.

One last thing that I miss in the moment is data for the "top" addon and the "light-colour". Would be nice to have them all in the properties in a single Point-Feature.

Maybe we need some specification how to transpile from the database to a style-friendly dataset. But I have to say I´m not shure if I can keep pace with you ;-)

BTW: Should we close this issue here  with a last link to your repo and only continue there? And

**Comment by aAXEe** [comment]
>Maybe we need some specification how to transpile from the database to a style-friendly dataset. But I have to say I´m not shure if I can keep pace with you ;-)

Yes we will need this .. and I am aware of this ;)
The current 1:1 mapping between tagging scheme to postgres tables and vector-layers is only intented as a first try to be able to find such issues ;)
Follow up: https://github.com/OpenSeaMap/vectortiles-generator/issues/5

> One last thing that I miss in the moment is data for the "top" addon and the "light-colour". Would be nice to have them all in the properties in a single Point-Feature.

I noticed that the osm-import does not fetch all data. It only takes the `seamark:type` object type from each node and parses all relevant attributes for that type. This way it ignores `seamark:light:*` if `seamark:type` is something different from `light`.

This should be addressed in https://github.com/OpenSeaMap/vectortiles-generator/issues/1

> BTW: Should we close this issue here with a last link to your repo and only continue there? And 

Yes I think so.
Having the seamark data as a separate data source and a style for that seems to make sense.
